lsm-tree.cabal

```diff
@@ -1,8 +1,117 @@
+
+  The worst-case in-memory size of an LSM-tree is \(O(n)\).
+
+  *   The worst-case size of the write buffer is \(O(1)\).
+
+      The maximum size of the write buffer on the write buffer allocation strategy, which is determined by the @'confWriteBufferAlloc'@ field of @'TableConfig'@.
+      Regardless of write buffer allocation strategy, the size of the write buffer may never exceed 4GiB.
+
+      [@AllocNumEntries maxEntries@]:
+        The maximum size of the write buffer is the maximum number of entries multiplied by the average size of a key–operation pair.
+
+  *   The worst-case size of the Bloom filters is \(O(n)\).
+
+      The total size of all Bloom filters depends on the Bloom filter allocation strategy, which is determined by the @'confBloomFilterAlloc'@ field of @'TableConfig'@.
+
+      [@AllocFixed bitsPerPhysicalEntry@]:
+          The total size of all Bloom filters is the number of bits per physical entry multiplied by the number of physical entries.
+      [@AllocRequestFPR requestedFPR@]:      
+          TODO: How does one determine the bloom filter size using @AllocRequestFPR@?
+
+  *   The worst-case size of the indexes is \(O(n)\).
+
+      The total size of all indexes depends on the index type, which is determined by the @'confFencePointerIndex'@ field of @'TableConfig'@.
+      The size of the various indexes is described in reference to the size of the database in [/memory pages/](https://en.wikipedia.org/wiki/Page_%28computer_memory%29).
+
+      [@OrdinaryIndex@]:
+          An ordinary index stores the maximum serialised key for each memory page.
+          The total size of all indexes is proportional to the average size of one serialised key per memory page.
+      [@CompactIndex@]:
+          A compact index stores the 64 most significant bits of the minimum serialised key for each memory page, as well as 1 bit per memory page to resolve clashes, 1 bit per memory page to mark overflow pages, and a negligable amount of memory for tie breakers.
+          The total size of all indexes is approximately 66 bits per memory page.
+
+  The total size of an LSM-tree must not exceed \(2^{41}\) physical entries.
+  Violation of this condition /is/ checked and will throw a 'TableTooLargeError'.
+
+  == Implementation
+
+  The implementation of LSM-trees in this package draws inspiration from:
+
+  *   Chris Okasaki.
+      1998.
+      \"Purely Functional Data Structures\"
+      [doi:10.1017/CBO9780511530104](https://doi.org/10.1017/CBO9780511530104)
+  *   Niv Dayan, Manos Athanassoulis, and Stratos Idreos.
+      2017.
+      \"Monkey: Optimal Navigable Key-Value Store.\"
+      [doi:10.1145/3035918.3064054](https://doi.org/10.1145/3035918.3064054)
+  *   Subhadeep Sarkar, Dimitris Staratzis, Ziehen Zhu, and Manos Athanassoulis.
+      2021.
+      \"Constructing and analyzing the LSM compaction design space.\"
+      [doi:10.14778/3476249.3476274](https://doi.org/10.14778/3476249.3476274)
```

Comment:
My preference is normally to put less in the `description`, and more into the `README`, so that the cabal file stays more concise and mostly deals with build information. The `README` is shipped with the package to users and is a natural place to look for more information. And it is rendered by Hackage too. What do you think?

Reply:
I'd written a script to extract the package description into the README. If the README is rendered by Hackage, then the alternative you're suggesting is fine. Do you have a link to the documentation for Hackage's rendering of Markdown?

Reply:
Since this feature isn't supported by Haddock, that means the documentation will not be rendered on any non-Hackage server. It also doesn't appear to be documented? Using rendering of Markdown from the README.md file also introduces a lot of difficulty in consistently generating links into the modules. Therefore, I think I'm opposed to this change, and would prefer to keep the documentation in the `description` field.

Reply:
There is no documentation for the rendering, but It's rendered in both Hackage and Stackage. No hyperrefs to modules is unfortunate, but even then the documentation would still be useful IMO

Maybe others can weigh in if they have preferences, but I won't cling too hard to putting the documentation in the package description

Reply:
For reference, I've added #661 which adds a script to generate the README from the package description.